mlkem/native/riscv64/src/rv64v_poly.c

```diff
@@ -0,0 +1,820 @@
+ * Name:        poly_ntt
+ *
+ * Description: Computes negacyclic number-theoretic transform (NTT) of
+ *              a polynomial in place;
+ *              inputs assumed to be in normal order, output in
+ *bitreversed order
+ *
+ * Arguments:   - uint16_t *r: pointer to in/output polynomial
+ **************************************************/
+
+/*  forward butterfly operation */
+
+#define MLK_RVV_BFLY_FX(u0, u1, ut, uc, vl) \
+  {                                         \
+    ut = fq_mul_vx(u1, uc, vl);             \
+    u1 = __riscv_vsub_vv_i16m1(u0, ut, vl); \
+    u0 = __riscv_vadd_vv_i16m1(u0, ut, vl); \
+  }
+
+#define MLK_RVV_BFLY_FV(u0, u1, ut, uc, vl) \
+  {                                         \
+    ut = fq_mul_vv(u1, uc, vl);             \
+    u1 = __riscv_vsub_vv_i16m1(u0, ut, vl); \
+    u0 = __riscv_vadd_vv_i16m1(u0, ut, vl); \
+  }
+
+static vint16m2_t mlk_rv64v_ntt2(vint16m2_t vp, vint16m1_t cz)
+{
+  size_t vl = MLK_RVV_E16M1_VL;
+  size_t vl2 = 2 * vl;
+
+  const vuint16m2_t v2p8 = bitswap_perm(3, 4, vl2);
+  const vuint16m2_t v2p4 = bitswap_perm(2, 4, vl2);
+  const vuint16m2_t v2p2 = bitswap_perm(1, 4, vl2);
+
+  /*    p1 = p8(p4(p2)) */
+  const vuint16m2_t v2p1 = __riscv_vrgather_vv_u16m2(
+      __riscv_vrgather_vv_u16m2(v2p2, v2p4, vl2), v2p8, vl2);
+
+  const vuint16m1_t vid = __riscv_vid_v_u16m1(vl);
+  const vuint16m1_t cs8 =
+      __riscv_vadd_vx_u16m1(__riscv_vsrl_vx_u16m1(vid, 3, vl), 2, vl);
+  const vuint16m1_t cs4 =
+      __riscv_vadd_vx_u16m1(__riscv_vsrl_vx_u16m1(vid, 2, vl), 2 + 2, vl);
+  const vuint16m1_t cs2 =
+      __riscv_vadd_vx_u16m1(__riscv_vsrl_vx_u16m1(vid, 1, vl), 2 + 2 + 4, vl);
+
+  vint16m1_t vt, c0, t0, t1;
+
+  /*    swap 8  */
+  vp = __riscv_vrgatherei16_vv_i16m2(vp, v2p8, vl2);
+  t0 = __riscv_vget_v_i16m2_i16m1(vp, 0);
+  t1 = __riscv_vget_v_i16m2_i16m1(vp, 1);
+
+  c0 = __riscv_vrgather_vv_i16m1(cz, cs8, vl);
+  MLK_RVV_BFLY_FV(t0, t1, vt, c0, vl);
+
+  /*    swap 4  */
+  vp = __riscv_vcreate_v_i16m1_i16m2(t0, t1);
+  vp = __riscv_vrgatherei16_vv_i16m2(vp, v2p4, vl2);
+  t0 = __riscv_vget_v_i16m2_i16m1(vp, 0);
+  t1 = __riscv_vget_v_i16m2_i16m1(vp, 1);
+
+  c0 = __riscv_vrgather_vv_i16m1(cz, cs4, vl);
+  MLK_RVV_BFLY_FV(t0, t1, vt, c0, vl);
+
+  /*    swap 2  */
+  vp = __riscv_vcreate_v_i16m1_i16m2(t0, t1);
+  vp = __riscv_vrgatherei16_vv_i16m2(vp, v2p2, vl2);
+  t0 = __riscv_vget_v_i16m2_i16m1(vp, 0);
+  t1 = __riscv_vget_v_i16m2_i16m1(vp, 1);
+
+  c0 = __riscv_vrgather_vv_i16m1(cz, cs2, vl);
+  MLK_RVV_BFLY_FV(t0, t1, vt, c0, vl);
+
+  /*    normalize   */
+  t0 = fq_mulq_vx(t0, MLK_MONT_R1, vl);
+  t1 = fq_mulq_vx(t1, MLK_MONT_R1, vl);
+
+  /*    reorganize  */
+  vp = __riscv_vcreate_v_i16m1_i16m2(t0, t1);
+  vp = __riscv_vrgatherei16_vv_i16m2(vp, v2p1, vl2);
+
+  return vp;
+}
+
+void mlk_rv64v_poly_ntt(int16_t *r)
+{
+  /* zetas can be compiled into vector constants; don't pass as a pointer */
+  /* check-magic: off */
+  const int16_t zeta[0x80] = {
```

Comment:
Those should ultimately be autogenerated via `autogen` similar to the other twiddles tables (e.g. `zetas.inc`). You will be able to copy-paste adjust most of it, I think -- from a cursory look, this is different from `zetas.inc` only in the order of the twiddles.

Do you have time to look into that, or shall me/Matthias do it as a follow-up?

Reply:
I will be returning to this code in a couple of weeks, and I don't mind if someone scripts them in the meanwhile. All three tables were generated with throw-away gp-pari statements but are  obvious to "reverse engineer" as you note; just a some combo of various orderings of roots-of-unity powers and Montgomery constants. Anyway, I understand what you're after here.